content/posts/2025-02-05-perf-improvements.md

```diff
@@ -0,0 +1,237 @@
+
+Previously, we have been doing this de-duplication in the user space. That is to say, our flows in kernel were identified not truly by their 5-tuple as mentioned previously, but by a larger [12-tuple](https://github.com/netobserv/netobserv-ebpf-agent/blob/c7c2acc6ac36c0c9966c67f668e29995ceb40066/bpf/types.h#L117-L137) that also included the direction (ingress/egress), the interface index, the MAC addresses (to be honest, I can't remember why the MAC addresses were there) and a couple of other fields. That resulted in flows being created in the BPF map for each different traversed interface, increasing the overall map cardinality. The flows were later de-duplicated in the user space, involving a temporary cache, at a cost of CPU and memory. Why did we do it that way? It's again related to the per-CPU map. De-duplication cannot be done properly in the kernel if the CPU processing a packet doesn't have the full picture of the related flow. It needs to know on which interface a packet was first seen, to decide whether to increase counters or not. The user space was the only place having that full picture.
+
+By removing the per-CPU map, we unlock this next improvement: shrinking the map key to a lighter [7-tuple](https://github.com/netobserv/netobserv-ebpf-agent/blob/2c96c420fc5ed223b7b99b00b2705fe84c5c0110/bpf/types.h#L159-L167), performing the de-duplication directly in-kernel, resulting in reducing the map size, the CPU cost to read it, and getting rid of the whole user-space mechanism for de-duplication. Why 7-tuples, by the way? Well, it's only for some ICMP related stuff — not something to worry about, performance-wise, as ICMP traffic is generally negligible (unless your cluster is full of `pings` running everywhere!) and has a smaller flow cardinality (no ports). The most important part was to remove the interface index from the key.
+
+One of the consequences is that we would lose information about each traversed interface. We want to still be able to tell: this sample flow went through `genev_sys_6081`, then `8aa5bd1d532fca8@if2`, and finally `ens5`. So we had to add this information as an array in the map values.
+
+Less flows to fetch thanks to smaller keys, and no user-space deduplication: less used CPU.
+
+**Fig. 1 (again): eBPF agent user-space CPU cores usage, averaged per node**
+
+![User-space CPU]({page.image('perf-improvements-1-8/user-space-cpu.png')})
+
+### Other improvements
+
+There have been other improvements identified, more as low-hanging fruits, but still impactful:
+
+- Removing some redundant calls to `bpf_map_update()`
+- Removing `packed` attribute from the eBPF structures
```

Comment:
nit: eBPF shared structures with userspace just to be specific and u can expand on the draw back of packed or include the link I shared during the review for reference